rspace/src/main/scala/coop/rchain/rspace/merger/instances/StateMergerImpl.scala

```diff
@@ -0,0 +1,322 @@
+    private def getJoins(
+        mainState: Blake2b256Hash,
+        mergedState: Blake2b256Hash,
+        channelHash: Blake2b256Hash
+    ) =
+      for {
+        maybeJoinHash <- historyRepo.getChannelHash(channelHash)
+        joinHash <- maybeJoinHash match {
+                     case Some(DataJoinHash(_, j)) => j.pure[F]
+                     case _ =>
+                       Sync[F].raiseError[Blake2b256Hash](
+                         new Exception(s"not found join hash ${channelHash} in channel store")
+                       )
+                   }
+        m <- joinMap.get
+        result <- m.get(joinHash) match {
+                   case None =>
+                     for {
+                       mainJoins    <- historyRepo.getJoins(mainState, joinHash)
+                       mergingJoins <- historyRepo.getJoins(mergedState, joinHash)
+                       unionJoins   = (mainJoins.toSet ++ mergingJoins.toSet).toSeq
+                       _            <- joinMap.update(_ + (joinHash -> unionJoins))
+                     } yield unionJoins
+                   case Some(j) => j.pure[F]
+                 }
+      } yield JoinInfo(channelHash, joinHash, result)
+
+    private def removeIndex[E](col: Seq[E], index: Int): Seq[E] = {
+      val (l1, l2) = col splitAt index
+      (l1 ++ (l2 tail))
+    }
+
+    private def removeJoins(
+        mainState: Blake2b256Hash,
+        mergedState: Blake2b256Hash,
+        channelHash: Blake2b256Hash,
+        join: Seq[C]
+    ) =
+      for {
+        joinInfo <- getJoins(mainState, mergedState, channelHash)
+        index    = joinInfo.joins.indexOf(join)
+        _ <- if (index != -1) {
+              joinMap.update { m =>
+                {
+                  val updatedJoins = removeIndex(joinInfo.joins, index)
+                  m + (joinInfo.joinHash -> updatedJoins)
+                }
+              }
+            } else Sync[F].unit
+      } yield ()
+
+    override def merge(
+        baseState: Blake2b256Hash,
+        mainState: Blake2b256Hash,
+        mergingState: Blake2b256Hash,
+        eventLogs: List[Event]
+    ): F[Blake2b256Hash] =
+      for {
+        changedChannels <- excludeEventsInComm(eventLogs).pure[F]
+        dataChanges <- changedChannels.produceChanges.toList
+                        .traverse(getDataChanges(_, mainState, baseState, mergingState))
+        consumeChangesOpt <- changedChannels.consumeChanges.toList
+                              .traverse(
+                                getContinuationChanges(_, mainState, baseState, mergingState)
+                              )
+        consumeChanges     = consumeChangesOpt.collect { case Some(c) => c }
+        joinChanges        <- getJoinChanges
+        changes            = dataChanges ++ consumeChanges ++ joinChanges
+        transformedActions = changes.map(transform)
+        historyActions     <- storeLeaves(transformedActions)
+        his                = history.reset(mainState)
```

Comment:
I'd call this `h`, its shorter, and `his` word has a meaning, so confusing a bit.